### PR TITLE
add model parallel design

### PR DESCRIPTION
**Pull request by typhoonzero**

### CODE REVIEW

File: doc/fluid/design/dist_train/distributed_model_parallel_pass.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+of the weight slices. Then each GPU can do the forward computation using their own part
+of the weight, then we use "AllGather" to concat all output from those 2 GPUs, then the following
+computation should be the same as the original program.
+
+## Implementaion
+
+The implementation should contain below components.
+
+1. Define an attribute for fc layer whether to enable model parallel (we can support only fc for now)
+1. IR pass can update "startup program"
```
Comment (Yancey0623) [inline doc/fluid/design/dist_train/distributed_model_parallel_pass.md]: Should explain why IR pass update the `startup_program`.

File: doc/fluid/design/dist_train/distributed_model_parallel_pass.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+1. Implement "AllGather" OpHandle
+1. Add an IR pass to update the graph, removing original fc layer ops and replace them with 
+   distributed ops like we described in above section
+1. Update "startup program" to initialize fc weights as column splited weights.
```
Comment (Yancey0623) [inline doc/fluid/design/dist_train/distributed_model_parallel_pass.md]: Another way to split fc weight:
1. Implement another fc layer, maybe use another name or an attribute.
1. The new FC layer implement:
    1. append splite weight op in startup_program to initialize the fc.weight by colume.
    1. append all_reduce_op->element_wise_mul_op->element_wise_add_op->all_gather_op
1. `multi_devices_pass` should convert the all_reduce_op and all_gather_op to all_reduce_op_handle and all_gather_op_handle, this approach is more general, this is easy to implement sync BN if the `multi_devices_pass` can deal with the NCCL op.

File: doc/fluid/design/dist_train/distributed_model_parallel_pass.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+on different devices (e.g. GPU) or nodes. When training the model, these layers can
+communicate with each other to do the computation as if this layer is on a single device.
+
+Note that this may not be the final design since we have to implement "AllReduce" and
+"AllGather" operators using "OpHandle" but not the original "Operator", this design can be
+further updated when we have a new framework skeleton.
+
+## Background
+
+For some classification models, we may have too many classes, like face recognization,
```
Comment (Yancey0623) [inline doc/fluid/design/dist_train/distributed_model_parallel_pass.md]: This paragraph is one part of the model parallelism, maybe we udpate the section as:

``` text
# DMMP
## Background
## Distribute one layer on multiple device.
### Data Flow
### Implement

## Distribute one layer on one device
... 
```

**Comment by shanyi15** [comment]
find the whole doc was deleted, will this PR close?

**Pull request closed without merging**